### PR TITLE
feat: create grouping accounts

### PR DESCRIPTION
**Pull request by obi-ewan**

### CODE REVIEW

File: programs/protocol_event/src/instructions/create_event.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+    event.category = category;
+    event.event_group = event_group;
+
+    event.active = false;
+
+    event.slug = event_info.slug;
+    event.name = event_info.name;
+
+    event.participants = event_info.participants;
```
Comment (eoin-betdex) [inline programs/protocol_event/src/instructions/create_event.rs:37]: I don't think we'd be able to breach it within a single transaction's size limit, but perhaps we should be checking participants length also.

Comment (eoin-betdex) [reply programs/protocol_event/src/instructions/create_event.rs:37]: I also wonder whether we should verify that the participant indices in the vec are at least within the bounds of `[0, category.participant_count)`?

File: programs/protocol_event/src/instructions/create_grouping.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
+use crate::error::EventError;
+use crate::state::category::Category;
+use crate::state::event_group::EventGroup;
+use anchor_lang::prelude::*;
+
+pub fn create_category(
+    category: &mut Category,
+    code: String,
+    name: String,
+    payer: Pubkey,
```
Comment (eoin-betdex) [inline programs/protocol_event/src/instructions/create_grouping.rs]: NP The pedant in me might try to align the order of arguments between `create_category` and `create_event_group` given the share 3 of them :twitch:

**Review by eoin-betdex — COMMENTED** [review]
A couple more comments but NPs tbh. Happy to approve if you're happy I'm being pedantic.

File: programs/protocol_event/src/instructions/create_event.rs (hunk 1 above)
Comment (obi-ewan) [reply programs/protocol_event/src/instructions/create_event.rs:37]: I've created a pattern of seperate `validate_x` methods for these, makes unit tests easier too

File: programs/protocol_event/src/instructions/create_event.rs [hunk 3]
```diff
@@ -49,22 +43,58 @@ pub fn create(
...
+    require!(
+        event_info.participants.len() <= Event::MAX_PARTICIPANTS,
+        EventError::MaxParticipantsExceeded,
+    );
+    require!(
+        event_info
+            .participants
+            .iter()
+            .all(|&participant| (1..=category_participant_count).contains(&participant)),
+        EventError::InvalidEventParticipants,
```
Comment (eoin-betdex) [inline programs/protocol_event/src/instructions/create_event.rs:64]: NP might be cheaper just to `participant < category_participant_count`

**Review by eoin-betdex — APPROVED** [review]

**Pull request merged**